clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -112,12 +126,34 @@ public ProduceRequest(Struct struct, short version) {
                 Struct partitionResponse = (Struct) partitionResponseObj;
                 int partition = partitionResponse.getInt(PARTITION_KEY_NAME);
                 MemoryRecords records = (MemoryRecords) partitionResponse.getRecords(RECORD_SET_KEY_NAME);
+
+                if (version >= 3)
+                    validateRecords(version, records);
+
                 partitionRecords.put(new TopicPartition(topic, partition), records);
             }
         }
         partitionSizes = createPartitionSizes(partitionRecords);
         acks = struct.getShort(ACKS_KEY_NAME);
         timeout = struct.getInt(TIMEOUT_KEY_NAME);
+        transactionalId = version >= 3 ? struct.getString(TRANSACTIONAL_ID_KEY_NAME) : null;
+    }
+
+    private void validateRecords(short version, MemoryRecords records) {
+        if (version >= 3) {
+            Iterator<RecordBatch.MutableRecordBatch> iterator = records.batches().iterator();
+            if (!iterator.hasNext())
+                throw new InvalidRecordException("Version 3 and above of the produce request contained no log entries");
+
+            RecordBatch.MutableRecordBatch entry = iterator.next();
+            if (entry.magic() != RecordBatch.MAGIC_VALUE_V2)
```

Comment:
If we do this check, should we not do it for other versions `MAGIC_VALUE_V1` too?